### PR TITLE
Add validate identifier protocol

### PR DESCRIPTION
**Pull request by armanio123**
Added endpoint ValidateIdentifier to allow IDE's to verify identifiers during certain operations (i.e renaming)

### CODE REVIEW

File: src/server/protocol.ts [hunk 1]
```diff
@@ -852,6 +854,29 @@ namespace ts.server.protocol {
...
+        /**
+         * Name of the identifier to validate.
+         */
+        identifierName: string;
+    }
+
+    /**
+     * Validate identifier requests; value of command
+     * is "ValidateIdentifier". Return response indicating
+     * the the identifier is valid or not.
```
Comment (andy-ms) [inline src/server/protocol.ts]: *whether the

File: src/services/types.ts [hunk 2]
```diff
@@ -608,6 +608,11 @@ namespace ts {
         triggerSpan: TextSpan;
     }
 
+    export interface IsValidIdentifierName {
+        isValid: boolean;
```
Comment (andy-ms) [inline src/services/types.ts]: Would it make more sense to just return `{ localizedErrorMessage: string | undefined }`? You would never return `{ isValid: true, localizedErrorMessage: "Nvm, here's an error message anyway" }`, right?

File: src/server/session.ts [hunk 3]
```diff
@@ -769,6 +769,22 @@ namespace ts.server {
             return project.getLanguageService().getRenameInfo(file, position);
         }
 
+        private isValidIdentifierName(args: protocol.ValidateIdentifierNameArgs): IsValidIdentifierName {
+            const project = this.getProject(args.projectFileName);
+
+            if (ts.isKeyword(stringToToken(args.identifierName)) || !ts.isIdentifierText(args.identifierName, project.getCompilerOptions().target)) {
```
Comment (andy-ms) [inline src/server/session.ts]: Already in `ts` namespace, so shouldn't need `ts.`

Comment (mhegazy) [inline src/server/session.ts]: what if the location of the rename allows for non-identifier names, e.g. inside a string literal?

File: src/server/protocol.ts (hunk 1 above)
Comment (armanio123) [reply src/server/protocol.ts]: Fixed.

File: src/services/types.ts (hunk 2 above)
Comment (armanio123) [reply src/services/types.ts]: Makes sense. Fixed

File: src/server/session.ts (hunk 3 above)
Comment (armanio123) [reply src/server/session.ts]: In that case the string will be reported as an error. In my opinion, the service client should be able to handle this scenarios. This method will only report if the string is a valid identifier or not regardless of it's location.

**Comment by mhegazy** [comment]
We have some identifier parsing logic in the managed side. we use that for detecting the completion range. i wonder why we do not do the same for this one instead of having a new command.

**Comment by armanio123** [comment]
I just checked the managed side and is fairly limited on what the TS server have already implemented. We'll need to duplicate some code if we want to achieve the same.

In any case, @billti and I discussed about implementing it on the managed side and decided not to go for it as it will add some duplication/maintainability that is not worth it. We agreed to rollback any changes for now and reopen if needed.

Thanks for the input.

**Pull request closed without merging**